src/connection.js

```diff
@@ -0,0 +1,282 @@
+import { EventEmitter } from 'events';
+import * as net from 'net';
+
+let connCounter = 0;
+
+/**
+ * @constant
+ * Initial state. Also set if a command fails or the connection is closed.
+ */
+const DO_NOTHING = 0;
+
+/**
+ * @constant
+ * After a command is executed, this will wait for the OKAY/FAIL response.
+ */
+const WAIT_FOR_COMMAND_RESULT = 1;
+
+/**
+ * @constant
+ * After a command executes and we have received the OKAY/FAIL, then we process
+ * whatever data is left. Certain commands, such as track-devices, keep sending
+ * more data that begins with the length of data expected.
+ */
+const WAIT_FOR_NEW_DATA = 2;
+
+/**
+ * @constant
+ * After a command is executed, this will wait for additional data until the
+ * connection is closed. This is for "adb shell" commands where the exact
+ * output length is unknown.
+ */
+const BUFFER_UNTIL_CLOSE = 3;
+
+/**
+ * Creates an Connection instance.
+ *
+ * @class
+ * @extends EventEmitter
+ * @classdesc Manages the connection and communcations with the ADB server.
+ * @constructor
+ * @param {ADB} adb - The ADB instance.
+ */
+export class Connection extends EventEmitter {
+	constructor(adb) {
+		super();
+		this.adb = adb;
+		this.port = adb && adb.opts && adb.opts.port || 5037;
+		this.socket = null;
+		this.state = DO_NOTHING;
+		this.connNum = ++connCounter;
+		this.dataEventName = 'data';
+		this.endEventName = 'end';
+	}
+
+	/**
+	 * Executes a command. If there is no connection to the ADB server, it will
+	 * connect to it, then run the command.
+	 *
+	 * @param {String} cmd - The command to run.
+	 * @param {Object} [opts] - Execute options.
+	 * @param {Boolean} [opts.bufferUntilClose=false] - Buffers all received data until ADB closes the connection.
+	 * @returns {Promise}
+	 * @access public
+	 */
+	exec(cmd, opts = {}) {
+		if (typeof cmd !== 'string' || !cmd) {
+			return Promise.reject(new TypeError('Expected command to be a string.'));
+		}
+
+		const dataEventName = this.dataEventName;
+		return new Promise((resolve, reject) => {
+			let conn = this;
```

Comment:
No need to assign `this` to a local variable. Just use `this.*` instead of `conn.*`.